### PR TITLE
CAP-0033 Clarifications and Fixes

### PR DESCRIPTION
**Pull request by jonjove**

### CODE REVIEW

File: core/cap-0033.md [hunk 1]
```diff
@@ -68,8 +68,14 @@ typedef AccountID* SponsorshipDescriptor;
 
 struct AccountEntryExtensionV2
 {
-    uint32 numSponsored;
-    uint32 numSponsoring;
+    uint32 numSponsored;  // Number of reserves sponsored by this account
+                          // This reduces the base reserve requirement
+    uint32 numSponsoring; // Number of reserves sponsored for this account
+                          // This increases the base reserve requirement
```
Comment (ire-and-curses) [inline core/cap-0033.md:74]: In our working group we said
>numSponsoring: how many (reserves) I’m sponsoring (plus sign in the base reserve formula)
numSponsored:  how many of my reserves are sponsored by someone else (minus sign in the base reserve formula)

This seems like the opposite of the comment in this PR? Which is correct?

Comment (leighmcculloch) [inline core/cap-0033.md]: I'm still very confused by these descriptions. I thought I understand until I read the second comment on each. Is `numSponsored` reducing the base reserve of this account or another account?

File: core/cap-0033.md [hunk 2]
```diff
@@ -603,6 +605,34 @@ ecosystem. Storing the sponsoring IDs for signers separately avoids this problem
...
+So how would a transfer work if `UpdateSponsorshipOp` were to obey the typical
+rules of sponsorship? Let `S1` be the current sponsor, `S2` be the new sponsor,
+`K` be the `LedgerKey` for an entry sponsored by `S1`, and `A` be the account
+which controls the entry identified by `K`. We would need a transaction with
+
+1. `S2` is source for `SponsorFutureReservesOp` with `sponsoredID = A`
+2. `S1` is source for `UpdateSponsorshipOp` with `ledgerKey = K`
+3. `A` is source for `ConfirmAndClearSponsorshipOp`
+
+But this clearly requires a signatures from the sponsored account, so if we are
```
Comment (ire-and-curses) [inline core/cap-0033.md:631]:
```suggestion
But this clearly requires a signature from the sponsored account, so if we are
```

File: core/cap-0033.md [hunk 3]
```diff
@@ -181,6 +189,7 @@ enum OperationType
 
 struct SponsorFutureReservesOp
 {
+    // Source account will sponsor future reserves for this account
```
Comment (leighmcculloch) [inline core/cap-0033.md:192]: This comment should lead with the field so that it is clear who the field is in the comment. Because it leads with the source account that is not in this struct it is easy to think the source account is the sponsoredID here.

This might be a bit wordy for you, but:
```suggestion
    // SponsoredID references the account for which future reserves will be sponsored by the source account of the operation this SponsoreFutureReservesOp is specified within.
```

File: core/cap-0033.md [hunk 4]
```diff
@@ -275,11 +290,12 @@ enum UpdateSponsorshipResultCode
...
 
     // codes considered as "failure" for the operation
-    UPDATE_SPONSORSHIP_DOES_NOT_EXIST = -1,
-    UPDATE_SPONSORSHIP_NOT_SPONSOR = -2,
-    UPDATE_SPONSORSHIP_LINE_FULL = -3,
-    UPDATE_SPONSORSHIP_LOW_RESERVE = -4,
-    UPDATE_SPONSORSHIP_ONLY_TRANSFERABLE = -5
+    UPDATE_SPONSORSHIP_DOES_NOT_EXIST = -1,   // specified entry does not exist
+    UPDATE_SPONSORSHIP_NOT_SPONSOR = -2,      // not sponsor of specified entry
+    UPDATE_SPONSORSHIP_LOW_RESERVE = -3,      // owner cannot afford this entry
```
Comment (leighmcculloch) [inline core/cap-0033.md:295]: The term owner is new. There's no field owner that I can see, can we reference who cannot afford the entry specifically?

File: core/cap-0033.md [hunk 5]
```diff
@@ -275,11 +290,12 @@ enum UpdateSponsorshipResultCode
...
+    UPDATE_SPONSORSHIP_ONLY_TRANSFERABLE = -4 // sponsorship of ClaimableBalance
+                                              // must be transferred to another
+                                              // account
```
Comment (leighmcculloch) [inline core/cap-0033.md:298]: Does this apply to only claimable balance? I would think this could happen for any sponsorship?

File: core/cap-0033.md [hunk 6]
```diff
@@ -375,7 +391,12 @@ If op.sponsoredID = op.sourceAccount
 The behavior of `SponsorFutureReservesOp op` is:
 ```
 If an account is sponsoring future reserves for op.sponsoredID
-    Fail with SPONSOR_FUTURE_RESERVES_ALREADY_EXISTS
+    Fail with SPONSOR_FUTURE_RESERVES_ALREADY_SPONSORED
```
Comment (leighmcculloch) [inline core/cap-0033.md:394]: This refers to already being sponsored just in the ephemeral sense inside this transaction, is that right?

File: core/cap-0033.md (hunk 1 above)
Comment (jonjove) [reply core/cap-0033.md]: The comments from the working group are correct. Thanks, fixed.

File: core/cap-0033.md (hunk 2 above)
Comment (jonjove) [reply core/cap-0033.md]: Good eye. Fixed.

File: core/cap-0033.md [hunk 7]
```diff
@@ -499,6 +522,10 @@ Similarly, `AccountMergeOp` will fail with `ACCOUNT_MERGE_IS_SPONSOR` if
 attempting to merge an account that has non-zero `numSponsoring`. This is
 required for sponsorship bookkeeping.
 
+#### CreateAccountOp
+`CreateAccountOp` will now be valid if `startingBalance >= 0`, whereas prior to
+this proposal it was valid if `startingBalance > 0`.
```
Comment (leighmcculloch) [inline core/cap-0033.md:527]: 💯 :shipit:

File: core/cap-0033.md [hunk 8]
```diff
@@ -651,10 +717,71 @@ operations[5]:
...
+operations[0]:
+    sourceAccount: S2
+    body:
+        type: SPONSOR_FUTURE_RESERVES
+        sponsoredID: S1
+operations[1]:
+    sourceAccount: S1
+    body:
+        type: UPDATE_SPONSORSHIP
+        ledgerKey: K
```
Comment (leighmcculloch) [inline core/cap-0033.md:765]: It's difficult from looking at these two example transactions in isolation to understand what the update sponsorship is doing. Would it be feasible to give these specific names:

- `REMOVE_SPONSORSHIP`
- `TRANSFER_SPONSORSHIP`

The term transfer is what is used in the document so it sounds like the right vocab to use for the operation.

File: core/cap-0033.md [hunk 9]
```diff
@@ -651,10 +717,71 @@ operations[5]:
...
+fee: <FEE>
+seqNum: <SEQ_NUM>
+timeBounds: <TIME_BOUNDS>
+memo:
+    type: MEMO_NONE
+operations[0]:
+    sourceAccount: S2
+    body:
+        type: SPONSOR_FUTURE_RESERVES
+        sponsoredID: S1
```
Comment (leighmcculloch) [inline core/cap-0033.md:760]: So when transferring sponsorship the sponsored account becomes the current sponsor? 🤯 This is very unintuitive to me.

File: core/cap-0033.md (hunk 1 above)
Comment (jonjove) [reply core/cap-0033.md]: Clarified, let me know if you think it can use more work. I really want to be precise about this.

File: core/cap-0033.md (hunk 3 above)
Comment (jonjove) [reply core/cap-0033.md]: Fixed, did something close to what you wrote. Thanks!

File: core/cap-0033.md (hunk 4 above)
Comment (jonjove) [reply core/cap-0033.md]: Clarified.

File: core/cap-0033.md (hunk 5 above)
Comment (jonjove) [reply core/cap-0033.md:300]: Right now this can only happen for `ClaimableBalanceEntry`. Basically, every type of entry other than claimable balance has a default sponsor (the account that controls that entry, which is the account that reflects that entry in the `numSubEntries`). But claimable balance has no default, it always has to have an explicit sponsor.

File: core/cap-0033.md (hunk 8 above)
Comment (jonjove) [reply core/cap-0033.md:767]: By "these" do you mean the operations? I agree that would clarify intentions, but it would lead to more errors because fundamentally remove and transfer are the same operation. So it is my opinion that adding additional restrictions on which one to use would just lead to people using the wrong one and having their transactions fail.

File: core/cap-0033.md [hunk 10]
```diff
@@ -569,6 +598,45 @@ Already there are discussions of ideas which may mitigate some of these issues
...
+signers, the sponsoring ID is stored separately. Why take different approaches?
+The reason is that `LedgerEntry` has an extension point but `Signer` does not.
+So if we wanted to store the sponsoring ID in the signer, then we would need to
+extend the `SignerKey` union. But this would force any downstream system that
+interacts with signers to update as well, introducing complexity throughout the
+ecosystem. Storing the sponsoring IDs for signers separately avoids this problem
+entirely, but it is perhaps slightly less elegant. Still, the trade-off is more
+than justified.
+
+### Why Doesn't UpdateSponsorshipOp Obey Typical Rules of Sponsorship?
```
Comment (abuiles) [inline core/cap-0033.md:612]: @jonjove which are the `Typical Rules of Sponsorship`?

**Comment by rice2000** [comment]
Looks like @jonjove addressed all the open questions.  Merging!
